### DISCUSSION TITLE
Idiomatic way to ignore whitespace

### DISCUSSION THREAD

**Discussion by hniksic** [winnow-rs/winnow#563] [Q&A]
I'm writing a parser for very limited boolean expressions, of the form `foo = 'literal' OR (bar = 'xxx' AND baz <> 'xxxxxxx')`.

I'd like the parser to ignore whitespaces. I implemented that by defining a whitespace-ignoring combinator:

```rust
fn space_around<I, O, E>(parser: impl Parser<I, O, E>) -> impl Parser<I, O, E>
where
    I: StreamIsPartial + Stream,
    E: ParserError<I>,
    I::Token: AsChar + Clone,
{
    delimited(multispace0, parser, multispace0)
}
```

I'm using that combinator in various places to ignore whitespace, e.g.:

```rust
fn comparison(i: &mut &str) -> PResult<Comparison> {
    let nm = space_around(name).parse_next(i)?;
    let eq = |i: &mut &str| {
        let (_, s) = ('=', space_around(string_literal)).parse_next(i)?;
        Ok(Comparison::Eq(Name(nm.into()), s.into()))
    };
    let ne = ...;
    let ge = ...;
    let le = ...;
    alt((eq, ne, ge, le)).parse_next(i)
}
```

I've already had bugs where I forgot to put `space_around()` somewhere, which I discovered in testing (or in unnit testing at best).

So my question is: is there a simpler, more idiomatic, or just a more consistent way to ignore whitespace?

**Comment by baoyachi** [comment]
use macro! ?

**Reply by hniksic** [reply]
Thanks for your answer. Are you referring to a specific macro, though, or are you referring to use of macros in general? I'm aware of macros and do use them (sparingly) in my Rust code, it's just not obvious how a macro would apply here.

**Reply by baoyachi** [reply]
@hniksic I'm referring to custom macros that simplify code declarations

**Reply by hniksic** [reply]
Thanks for the clarification. In my case the declarations seem simple enough, so it's not obvious that a macro would help. I was wondering if there was a feature provided by winnow to apply them consistently, and how others approached this issue in their DSL parsers. If there is an example of the kind of macro that you're suggesting that I could take a look at, it'd be much appreciated.

**Comment by epage** [comment]
Do you have a gramma? When I did `toml`, every parser was documented with the bnf it was associated with. This made it easy to tell I covered every case.

btw, putting your white space parser inside the operator `alt` likely wouldn't line up with a grammar and will be slower as it has to re-parse tke whitespace on each operator. Would also be good to use `dispatch!` on the operators.

**Reply by hniksic** [reply]
>Do you have a gramma?

I don't have a formal grammar, as I'm parsing expressions provided by a third party, working off examples. I did write a grammar to aid my own thinking, but it's not proper BNF, and it obviously omits whitespace handling. Such simplified grammar looks like this:

```
        // string: "'" [^']* "'"
        // int: [0-9]+
        // comparison: prop "=" string | prop "<>" string | prop ">=" int | prop "<=" int
        // expr1: comparison | "(" expr ")"
        // expr: expr1 ( ("OR" | "AND") expr1 )*
```

(`expr1` is split off from `expr` to eliminate left recursion, which was quite a headache at first. I was immensely helped by [this article by matklad](https://matklad.github.io/2020/04/13/simple-but-powerful-pratt-parsing.html#Recursive-descent-and-left-recursion), specifically the observation that in a hand-written parser you could cheat by implementing a loop manually. That's exactly what winnow's `arithmetic` example does as well, so I used it as template, and it worked beautifully.)

>btw, putting your white space parser inside the operator `alt` likely wouldn't line up with a grammar and will be slower as it has to re-parse tke whitespace on each operator.

I think I see what you mean, but I'm not sure how to do it better. Ideally I'd like to instruct the parser to ignore whitespace, [kind of like this](https://lark-parser.readthedocs.io/en/stable/grammar.html#ignore), but that approach doesn't seem quite compatible with the philosophy of winnow.

>Would also be good to use `dispatch!` on the operators.

`dispatch!` looks interesting, but it might even be overkill in my case. The current naive parser parses 80k expressions per second, which is more than good enough for me.
